README.md
# ImpermanentGain
The antiparticle of impermanent loss

## Commands

```shell
npx hardhat accounts
npx hardhat compile
npx hardhat clean
npx hardhat test
npx hardhat node
npx hardhat help
REPORT_GAS=true npx hardhat test
npx hardhat coverage
npx hardhat run scripts/deploy.ts
TS_NODE_FILES=true npx ts-node scripts/deploy.ts
npx eslint '**/*.{js,ts}'
npx eslint '**/*.{js,ts}' --fix
npx prettier '**/*.{json,sol,md}' --check
npx prettier '**/*.{json,sol,md}' --write
npx solhint 'contracts/**/*.sol'
npx solhint 'contracts/**/*.sol' --fix
```

# Etherscan verification

To try out Etherscan verification, you first need to deploy a contract to an Ethereum network that's supported by Etherscan, such as Ropsten.

In this project, copy the .env.example file to a file named .env, and then edit it to fill in the details. Enter your Etherscan API key, your Ropsten node URL (eg from Alchemy), and the private key of the account which will send the deployment transaction. With a valid .env file in place, first deploy your contract:

```shell
hardhat run --network ropsten scripts/sample-script.ts
```

Then, copy the deployment address and paste it in to replace `DEPLOYED_CONTRACT_ADDRESS` in this command:

```shell
npx hardhat verify --network ropsten DEPLOYED_CONTRACT_ADDRESS "Hello, Hardhat!"
```

# Performance optimizations

For faster runs of your tests and scripts, consider skipping ts-node's type checking by setting the environment variable `TS_NODE_TRANSPILE_ONLY` to `1` in hardhat's environment. For more details see [the documentation](https://hardhat.org/guides/typescript.html#performance-optimizations).

# Remix Setup
1. Run `npm run serve`  
This will start `remixd` that installed with `npm install`
2. Open [Remix IDE](https://remix.ethereum.org/)
3. Activate `Remixd` plugin, then you can access codes now
4. It's recommand not to check `Auto Compile` and `Enable Hardhat Compilation` at the same time
5. If errors saying writing conflicts on metadatas keep popping, turn off `Generate contract metadata` in Setting may help

# Local node setup
1. Create or modify `.env`, set `FORKING_URL` to a endpoint
2. Run `npx hardhat node`
3. Now you can connect to localhost:8545 and that will fork the state of mainnet
4. You can activate `Hardhat Provider` plugin in remix and select `Hardhat Provider` in Environment to deploy
5. To lock the forking block, umcomment the `blockNumber` config in `hardhat.config.ts` and change it

CHANGELOG.md
# Changelog

## V1 to V2 transistion:

- upgrade Solidity version from `0.5.17` to `0.8.7`
- abstract template with basic functions
- tokenize long/short position into ERC20 token
- add `mintExactA/B` function for better UX
- better token naming

## TokenFactory

A factory that deploys [EIP1167](https://eips.ethereum.org/EIPS/eip-1167) minimal proxy of mintable ERC20 token

## iGain V2 instances

### IL
The V2 version of iGain V1, almost identical except for a configurable leverage

### AAVE IRS
Tokenized incurred borrow interest of AAVE within a given period

### Yearn IRS
Tokenized incurred yield of Yearn within a given period

### Delta
Tokenized price movement by a sigmoid function


test/utils/index.ts
export { default as getSigner } from "./getSigner";
export { default as getRevertError } from "./getRevertError";


test/IGainIL.test.ts
import { expect, use } from "chai";
import { ethers, network } from "hardhat";
import { BigNumber } from "ethers";
import chaiAsPromised from "chai-as-promised";
import { ERC20Mintable, IGainIL } from "../typechain";
import { getRevertError, getSigner } from "./utils";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import { parseUnits } from "@ethersproject/units";
import { getAddress } from "@ethersproject/address";
import BN from "bignumber.js";

use(chaiAsPromised);
const e18 = BigNumber.from("10").pow(18);

function sqrt(value: BigNumber): BigNumber {
  return BigNumber.from(
    new BN(value.toString()).sqrt().integerValue(BN.ROUND_FLOOR).toString(10)
  );
}

function getFee(
  openTime: BigNumber,
  closeTime: BigNumber,
  txTime: BigNumber,
  minFee: BigNumber,
  maxFee: BigNumber
): BigNumber {
  if (closeTime.lte(openTime)) return e18.sub(minFee);
  return e18.sub(
    minFee.add(
      maxFee.sub(minFee).mul(txTime.sub(openTime)).div(closeTime.sub(openTime))
    )
  );
}

describe("IGainIL", function () {
  const amount = BigNumber.from(parseUnits("10000"));
  let IGainIL: IGainIL;
  let accounts: SignerWithAddress[];
  let operator: SignerWithAddress;
  let base: ERC20Mintable;
  let aContract: ERC20Mintable;
  let bContract: ERC20Mintable;

  // Init configs
  const baseToken = getAddress("0x6B175474E89094C44Da98b954EedeAC495271d0F");
  const oracle = getAddress("0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419");
  const treasury = getAddress("0x83D0D842e6DB3B020f384a2af11bD14787BEC8E7");
  const batchName = "IL-ETHUSD";
  const leverage = BigNumber.from("10000000000000000000");
  const duration = BigNumber.from("86400");
  const a = BigNumber.from("200000000000000000");
  const b = BigNumber.from("1000000000000000000");

  before(async function () {
    accounts = await ethers.getSigners();
    operator = accounts[0];

    const IGainILDeployer = await ethers.getContractFactory("IGainIL");
    IGainIL = await IGainILDeployer.deploy();
    await IGainIL.deployed();

    base = await ethers.getContractAt(
      "ERC20Mintable",
      "0x6B175474E89094C44Da98b954EedeAC495271d0F" // DAI
    );
    const basePool = base.connect(
      await getSigner("0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7") // Curve Pool
    );
    await basePool.transfer(operator.address, amount);
    await base.approve(IGainIL.address, amount);
  });

  describe("Init", async function () {
    it("Should initable", async function () {
      await IGainIL.init(
        baseToken,
        oracle,
        treasury,
        batchName,
        leverage,
        duration,
        a,
        b
      );

      const [
        resultBaseToken,
        resultOracle,
        resultTreasury,
        resultLeverage,
        resultOpenTime,
        resultCloseTime,
        resultA,
        resultB,
      ] = await Promise.all([
        IGainIL.baseToken(),
        IGainIL.oracle(),
        IGainIL.treasury(),
        IGainIL.leverage(),
        IGainIL.openTime(),
        IGainIL.closeTime(),
        IGainIL.a(),
        IGainIL.b(),
      ]);

      expect(baseToken).equal(resultBaseToken);
      expect(oracle).equal(resultOracle);
      expect(treasury).equal(resultTreasury);
      expect(leverage).equal(resultLeverage);
      expect(duration).equal(resultCloseTime.sub(resultOpenTime));

      aContract = await ethers.getContractAt("ERC20Mintable", resultA);
      bContract = await ethers.getContractAt("ERC20Mintable", resultB);

      expect("iGain A token " + batchName).equal(await aContract.name());
      expect("iGain B token " + batchName).equal(await bContract.name());
      expect("iG-A " + batchName).equal(await aContract.symbol());
      expect("iG-B " + batchName).equal(await bContract.symbol());
      expect(await base.decimals()).equal(await aContract.decimals());
      expect(await base.decimals()).equal(await bContract.decimals());
    });

    it("Should not re-initable", async function () {
      await expect(
        IGainIL.init(
          baseToken,
          oracle,
          treasury,
          batchName,
          leverage,
          duration,
          a,
          b
        )
      ).eventually.be.rejected;
    });
  });

  describe("Before close time", async function () {
    it("Fee should correct", async function () {
      const [openTime, closeTime, minFee, maxFee] = await Promise.all([
        IGainIL.openTime(),
        IGainIL.closeTime(),
        IGainIL.minFee(),
        IGainIL.maxFee(),
      ]);
      const txTime = openTime.add(10);
      const expectedFee = getFee(openTime, closeTime, txTime, minFee, maxFee);

      await network.provider.send("evm_setNextBlockTimestamp", [
        txTime.toNumber(),
      ]);
      await network.provider.send("evm_mine");
      const realFee = await IGainIL.fee();
      expect(expectedFee).equal(realFee);
    });

    it("Should not closable", async function () {
      await expect(IGainIL.close()).eventually.be.rejectedWith("Not yet");
    });
  });

  describe("After close time", async function () {
    before(async function () {
      await network.provider.send("evm_increaseTime", [86400]);
      await network.provider.send("evm_mine");
    });

    it("Fee should be max", async function () {
      const [fee, maxFee] = await Promise.all([
        IGainIL.fee(),
        IGainIL.maxFee(),
      ]);

      expect(e18.sub(fee)).equal(maxFee);
    });

    it("Should closable after duration", async function () {
      await expect(IGainIL.close()).eventually.be.fulfilled;
    });

    it("Should re-closable after closed", async function () {
      await expect(IGainIL.close()).eventually.be.rejectedWith(
        getRevertError("Closed")
      );
    });
  });

  describe("Calc IL", async function () {
    it("Should correct", async function () {
      const ratio = BigNumber.from(120).mul(BigNumber.from(10).pow(16)); // 1.2 * 1e18
      const result = await IGainIL.calcIL(ratio);

      expect(result).equal(
        e18.sub(sqrt(ratio.mul(e18)).mul(e18.mul(2)).div(ratio.add(e18)))
      );
    });
  });
});


test/IGainIRS.test.ts
import { expect, use } from "chai";
import { ethers, network } from "hardhat";
import { BigNumber } from "ethers";
import chaiAsPromised from "chai-as-promised";
import { ERC20Mintable, IGainYearnIRS } from "../typechain";
import { getRevertError, getSigner } from "./utils";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import { parseUnits } from "@ethersproject/units";
import { getAddress } from "@ethersproject/address";
import BN from "bignumber.js";

use(chaiAsPromised);
const e18 = BigNumber.from("10").pow(18);

function sqrt(value: BigNumber): BigNumber {
  return BigNumber.from(
    new BN(value.toString()).sqrt().integerValue(BN.ROUND_FLOOR).toString(10)
  );
}

function getFee(
  openTime: BigNumber,
  closeTime: BigNumber,
  txTime: BigNumber,
  minFee: BigNumber,
  maxFee: BigNumber
): BigNumber {
  if (closeTime.lte(openTime)) return e18.sub(minFee);
  return e18.sub(
    maxFee.sub(
      maxFee.sub(minFee).mul(txTime.sub(openTime)).div(closeTime.sub(openTime))
    )
  );
}

describe("IGainYearnIRS", function () {
  const amount = BigNumber.from(parseUnits("10000"));
  let IGainYearnIRS: IGainYearnIRS;
  let accounts: SignerWithAddress[];
  let operator: SignerWithAddress;
  let base: ERC20Mintable;
  let aContract: ERC20Mintable;
  let bContract: ERC20Mintable;

  // Init configs
  const baseToken = getAddress("0x6B175474E89094C44Da98b954EedeAC495271d0F");
  const vault = getAddress("0xdA816459F1AB5631232FE5e97a05BBBb94970c95");
  const treasury = getAddress("0x83D0D842e6DB3B020f384a2af11bD14787BEC8E7");
  const batchName = "IRS-yvDAI";
  const leverage = BigNumber.from("5000000000000000000");
  const duration = BigNumber.from("86400");
  const a = BigNumber.from("1000000000000000000");
  const b = BigNumber.from("1000000000000000000");

  before(async function () {
    accounts = await ethers.getSigners();
    operator = accounts[0];

    const IGainYearnIRSDeployer = await ethers.getContractFactory(
      "IGainYearnIRS"
    );
    IGainYearnIRS = await IGainYearnIRSDeployer.deploy();
    await IGainYearnIRS.deployed();

    base = await ethers.getContractAt(
      "ERC20Mintable",
      "0x6B175474E89094C44Da98b954EedeAC495271d0F" // DAI
    );
    const basePool = base.connect(
      await getSigner("0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7") // Curve Pool
    );
    await basePool.transfer(operator.address, amount);
    await base.approve(IGainYearnIRS.address, amount);
  });

  describe("Init", async function () {
    it("Should initable", async function () {
      await IGainYearnIRS.init(
        baseToken,
        vault,
        treasury,
        batchName,
        leverage,
        duration,
        a,
        b
      );

      const [
        resultBaseToken,
        resultVault,
        resultTreasury,
        resultLeverage,
        resultOpenTime,
        resultCloseTime,
        resultA,
        resultB,
      ] = await Promise.all([
        IGainYearnIRS.baseToken(),
        IGainYearnIRS.vault(),
        IGainYearnIRS.treasury(),
        IGainYearnIRS.leverage(),
        IGainYearnIRS.openTime(),
        IGainYearnIRS.closeTime(),
        IGainYearnIRS.a(),
        IGainYearnIRS.b(),
      ]);

      expect(baseToken).equal(resultBaseToken);
      expect(vault).equal(resultVault);
      expect(treasury).equal(resultTreasury);
      expect(leverage).equal(resultLeverage);
      expect(duration).equal(resultCloseTime.sub(resultOpenTime));

      aContract = await ethers.getContractAt("ERC20Mintable", resultA);
      bContract = await ethers.getContractAt("ERC20Mintable", resultB);

      expect("iGain A token " + batchName).equal(await aContract.name());
      expect("iGain B token " + batchName).equal(await bContract.name());
      expect("iG-A " + batchName).equal(await aContract.symbol());
      expect("iG-B " + batchName).equal(await bContract.symbol());
      expect(await base.decimals()).equal(await aContract.decimals());
      expect(await base.decimals()).equal(await bContract.decimals());
    });

    it("Should not re-initable", async function () {
      await expect(
        IGainYearnIRS.init(
          baseToken,
          vault,
          treasury,
          batchName,
          leverage,
          duration,
          a,
          b
        )
      ).eventually.be.rejected;
    });
  });

  describe("Before close time", async function () {
    it("Fee should correct", async function () {
      const [openTime, closeTime, minFee, maxFee] = await Promise.all([
        IGainYearnIRS.openTime(),
        IGainYearnIRS.closeTime(),
        IGainYearnIRS.minFee(),
        IGainYearnIRS.maxFee(),
      ]);
      const txTime = openTime.add(10);
      const expectedFee = getFee(openTime, closeTime, txTime, minFee, maxFee);

      await network.provider.send("evm_setNextBlockTimestamp", [
        txTime.toNumber(),
      ]);
      await network.provider.send("evm_mine");
      const realFee = await IGainYearnIRS.fee();
      expect(expectedFee).equal(realFee);
    });

    it("Should not closable", async function () {
      await expect(IGainYearnIRS.close()).eventually.be.rejectedWith("Not yet");
    });
  });

  describe("After close time", async function () {
    before(async function () {
      await network.provider.send("evm_increaseTime", [86400]);
      await network.provider.send("evm_mine");
    });

    it("Fee should be min", async function () {
      const [fee, minFee] = await Promise.all([
        IGainYearnIRS.fee(),
        IGainYearnIRS.minFee(),
      ]);

      expect(e18.sub(fee)).equal(minFee);
    });

    it("Should closable after duration", async function () {
      await expect(IGainYearnIRS.close()).eventually.be.fulfilled;
    });

    it("Should re-closable after closed", async function () {
      await expect(IGainYearnIRS.close()).eventually.be.rejectedWith(
        getRevertError("Closed")
      );
    });
  });
});


hardhat.config.ts
// Run tsc to genenrate js version for remix
import * as dotenv from "dotenv";

import { HardhatUserConfig, task } from "hardhat/config";
import "@nomiclabs/hardhat-etherscan";
import "@nomiclabs/hardhat-waffle";
import "@typechain/hardhat";
import "hardhat-gas-reporter";
import "solidity-coverage";

dotenv.config();

// This is a sample Hardhat task. To learn how to create your own go to
// https://hardhat.org/guides/create-task.html
task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
  const accounts = await hre.ethers.getSigners();

  for (const account of accounts) {
    console.log(account.address);
  }
});

// You need to export an object to set up your config
// Go to https://hardhat.org/config/ to learn more

const config: HardhatUserConfig = {
  solidity: "0.8.7",
  networks: {
    hardhat: {
      forking: {
        url: process.env.FORKING_URL || "",
        blockNumber: parseInt(process.env.BLOCK_NUMBER || "13464910"),
      },
      allowUnlimitedContractSize: true,
    },
    ropsten: {
      url: process.env.ROPSTEN_URL || "",
      accounts:
        process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
  },
  gasReporter: {
    enabled: process.env.REPORT_GAS !== undefined,
    currency: "USD",
  },
  etherscan: {
    apiKey: process.env.ETHERSCAN_API_KEY,
  },
  mocha: {
    timeout: 200000,
  },
};

export default config;


hardhat.config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
// Run tsc to genenrate js version for remix
var dotenv = __importStar(require("dotenv"));
var config_1 = require("hardhat/config");
require("@nomiclabs/hardhat-etherscan");
require("@nomiclabs/hardhat-waffle");
require("@typechain/hardhat");
require("hardhat-gas-reporter");
require("solidity-coverage");
dotenv.config();
// This is a sample Hardhat task. To learn how to create your own go to
// https://hardhat.org/guides/create-task.html
(0, config_1.task)("accounts", "Prints the list of accounts", function (taskArgs, hre) { return __awaiter(void 0, void 0, void 0, function () {
    var accounts, _i, accounts_1, account;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, hre.ethers.getSigners()];
            case 1:
                accounts = _a.sent();
                for (_i = 0, accounts_1 = accounts; _i < accounts_1.length; _i++) {
                    account = accounts_1[_i];
                    console.log(account.address);
                }
                return [2 /*return*/];
        }
    });
}); });
// You need to export an object to set up your config
// Go to https://hardhat.org/config/ to learn more
var config = {
    solidity: "0.8.7",
    networks: {
        hardhat: {
            forking: {
                url: process.env.FORKING_URL || "",
                blockNumber: parseInt(process.env.BLOCK_NUMBER || "13464910")
            },
            allowUnlimitedContractSize: true
        },
        ropsten: {
            url: process.env.ROPSTEN_URL || "",
            accounts: process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : []
        }
    },
    gasReporter: {
        enabled: process.env.REPORT_GAS !== undefined,
        currency: "USD"
    },
    etherscan: {
        apiKey: process.env.ETHERSCAN_API_KEY
    },
    mocha: {
        timeout: 200000
    }
};
exports["default"] = config;
